File: tail/src/fib_tail.py
import sys


def fib(n):
    def fibtail(n, second, first, res):
        if n == 1:
            return res
        return fibtail(n - 1, first, res, first + res)
    return fibtail(n, 0, 0, 1)

if __name__ == '__main__':
    limit = 1000000000
    sys.setrecursionlimit(limit)

    n = int(sys.argv[1])
    print("res = {}".format(fib(n)))

File: tail/src/fib_tail.hs
import System.Environment

fib :: Int -> Integer
fib n = fib_tail n 0 1

fib_tail :: Int -> Integer -> Integer -> Integer
fib_tail n second first | n < 3 = second + first
fib_tail n second first = fib_tail (n - 1) first (second + first)

main = do
    args <- getArgs
    let n = read (args !! 0) :: Int
    print(fib n)
